### PR TITLE
do not stop oracle if ethclient disconnects

### PR DESCRIPTION
**Pull request by jordipainan**

### CODE REVIEW

**Comment by jordipainan** [comment]
closes #98

File: chain/chain.go [hunk 1]
```diff
@@ -413,10 +408,10 @@ func (e *EthChainContext) SyncGuard(ctx context.Context) {
 			log.Warn("ethereum is experiencing sync problems, restarting node...")
 			e.RestartLock.Lock()
 			if err = e.Node.Close(); err != nil {
-				log.Fatal(err)
+				log.Error(err)
 			}
 			if err = e.Node.Start(); err != nil {
-				log.Fatal(err)
+				log.Error(err)
```
Comment (p4u) [inline chain/chain.go]: This is not related to an external Web3 connection, this case should be a Fatal

File: chain/chain.go [hunk 2]
```diff
@@ -248,26 +248,24 @@ func (e *EthChainContext) Start() {
...
-		if err != nil || client == nil {
-			log.Fatalf("cannot create a client connection: (%s), tried %d times.", err, types.EthereumDialMaxRetry)
-		}
+
 		tctx, cancel := context.WithTimeout(context.Background(), types.EthereumReadTimeout)
 		defer cancel()
 		nid, err := client.NetworkID(tctx)
 		if err != nil || nid == nil {
-			log.Fatalf("cannot get network ID from external web3: (%s)", err)
+			log.Errorf("cannot get network ID from external web3: (%s)", err)
```
Comment (p4u) [inline chain/chain.go]: log.Error() does not stop the execution. It looks like this is not taken into account on this PR. Make the proper flow modifications to handle Errors instead of Fatals correctly.

**Review by p4u — CHANGES_REQUESTED** [review]

File: chain/chain.go (hunk 2 above)
Comment (jordipainan) [reply chain/chain.go:265]: this is the case where the node is not synced  and we try to restart it but fails on close or fails on restart. Do we want to try again the restart or the close ? Or do we want to stop all the eth node operations without trying again these operations?

**Comment by jordipainan** [comment]
Behavior:
- If starting node from scratch:
    - If internal node fails: Fatal
    - If external node fails: Error and retry 10 times, if still not working, init all other components and start ethevents only once the client connection is established
- If node already running:
    - Warn user and keep trying to reconnect without affecting other components like keykeeper
    - Once the client is connected again resume the eth events subscription

File: chain/ethevents/events.go [hunk 3]
```diff
@@ -235,55 +235,116 @@ func (ev *EthereumEvents) AddEventHandler(h EventHandler) {
...
+			}
+			lastBlock = blk.Number().Int64()
+			break
 		}
 		// For security, read also the new passed blocks before subscribing
-		ev.processEventLogsFromTo(ctx, startBlock, blk.Number().Int64(), client)
+		// Same as the last processEventLogsFromTo function call
+		if err := ev.processEventLogsFromTo(ctx, lastBlock, blk.Number().Int64(), client); err != nil {
+			log.Errorf("cannot process event logs from block %d to block %d with error: %s", lastBlock, blk.Number().Int64(), err)
+
```
Comment (p4u) [inline chain/ethevents/events.go]: empty line

File: chain/chain.go [hunk 4]
```diff
@@ -198,7 +198,7 @@ func (e *EthChainContext) Start() {
 
 	if len(e.Keys.Accounts()) < 1 {
 		if err := e.createAccount(); err != nil {
-			log.Error(err)
+			log.Errorf("cannot create ethereum account: %s", err)
```
Comment (p4u) [inline chain/chain.go]: this looks like a Fatal

File: chain/chain.go [hunk 5]
```diff
@@ -422,3 +404,24 @@ func (e *EthChainContext) SyncGuard(ctx context.Context) {
...
+func EthClientConnect(dialAddr string, attempts int) (*ethclient.Client, error) {
+	var attemptsCounter int
+	for {
+		if attempts != 0 && attemptsCounter == attempts {
+			return nil, fmt.Errorf("tryied %d to reconnect, connection cannot be established", attemptsCounter)
+		}
+		client, err := ethclient.DialContext(context.Background(), dialAddr)
+		if err != nil || client == nil {
+			log.Errorf("cannot create a client connection: (%s), trying again ...", err)
+			time.Sleep(time.Second * 30)
```
Comment (p4u) [inline chain/chain.go]: 30 seconds looks too much

File: chain/contractWrappers.go [hunk 6]
```diff
@@ -59,18 +59,15 @@ func NewVotingHandle(contractsAddress []common.Address, dialEndpoint string) (*V
...
 	ph := new(VotingHandle)
 	// try connect to the w3endpoint using an RPC connection
-	for i := 0; i < types.EthereumDialMaxRetry; i++ {
+	for {
 		ph.EthereumRPC, err = ethrpc.Dial(dialEndpoint)
 		if err != nil || ph.EthereumRPC == nil {
-			log.Warnf("cannot create a ethereum rpc connection: (%s), trying again (%d of %d)", err, i+1, types.EthereumDialMaxRetry)
-			time.Sleep(time.Second * 2)
+			log.Warnf("cannot create a ethereum rpc connection: (%s), trying again ...", err)
+			time.Sleep(time.Second * 30)
```
Comment (p4u) [inline chain/contractWrappers.go]: Same...

**Review by p4u — CHANGES_REQUESTED** [review]

File: chain/chain.go (hunk 4 above)
Comment (jordipainan) [reply chain/chain.go]: Done, also capture the error on the account unlock. Fatal if cannot unlock account

File: chain/chain.go (hunk 5 above)
Comment (jordipainan) [reply chain/chain.go]: Reduced to 10

File: chain/contractWrappers.go (hunk 6 above)
Comment (jordipainan) [reply chain/contractWrappers.go]: Also reduced to 10
